lib/dal/src/workspace_snapshot/migrator.rs

```diff
@@ -77,95 +83,105 @@ impl SnapshotGraphMigrator {
-            ctx_after_migration.set_workspace_snapshot(migrated_snapshot);
-
-            change_set
-                .update_pointer(&ctx_after_migration, new_snapshot_address)
+
+                // NOTE(victor): The context that gets passed in does not have a workspace snapshot
+                // on it, since its main purpose is to allow access to the services context.
+                // We need to create a context for each migrated changeset here to run operations
+                // that depend on the graph
+                let mut ctx_after_migration =
+                    ctx.clone_with_new_visibility(Visibility::from(change_set.id));
+                // TODO make sure that when we clone with a changeset id we also set changeset
+                // (or there's no clone anymore and we always change it via following method)
+                ctx_after_migration.set_change_set(change_set.clone())?;
+
+                let snapshot_address = change_set.workspace_snapshot_address;
+
+                let new_snapshot_address = match migration_map.get(&snapshot_address) {
+                    Some(cached_addr) => *cached_addr,
+                    None => match self
+                        .migrate_snapshot(&ctx_after_migration, snapshot_address)
+                        .await
+                    {
+                        Ok(addr) => addr,
+                        Err(err) => {
+                            let err_string = err.to_string();
+                            if err_string.contains("missing from store for node")
+                                || err_string
+                                    .contains("workspace snapshot graph missing at address")
+                            {
+                                error!(
+                                    error = ?err,
+                                    "Migration error: {err_string}, marking change set {} for workspace {:?} as failed",
+                                    change_set.id, change_set.workspace_id
+                                );
+
+                                change_set
+                                    .update_status(ctx, ChangeSetStatus::Failed)
+                                    .await?;
+                                continue;
+                            } else {
+                                return Err(err)?;
+                            }
+                        }
+                    },
+                };
+
+                let migrated_snapshot =
+                    SplitSnapshot::find(&ctx_after_migration, new_snapshot_address).await?;
+                ctx_after_migration.set_workspace_split_snapshot(migrated_snapshot);
+
+                change_set
+                    .update_pointer(&ctx_after_migration, new_snapshot_address)
+                    .await?;
+
+                migration_map.insert(snapshot_address, new_snapshot_address);
+                change_set_count += 1;
+            }
+
+            workspace
```

Comment:
I'm a little lost as to where the transaction is here--I was expecting to see a ctx.commit() at some point, but maybe we just don't start a transaction? If we're not starting a transaction, does that mean that if we fail to migrate one of the changesets in a workspace, we'll update their pointer but not update the workspace's snapshot version? Would that cause failures to read the snapshots later?